File: pkg/R/krylov.R
## ID: krylov.R, last updated 2022-08-01, F.Osorio

krylov <- function(a, b, m = ncol(a))
{ ## construcs the Krylov matrix based on an n-by-n matrix a and an n-vector b

  ## validating arguments
  if (is.data.frame(a))
    a <- as.matrix(a)
  if (!is.matrix(a))
    stop("supply a matrix-like 'a'")
  if (!is.numeric(a))
    stop("argument 'a' is not a numeric matrix")
  da <- dim(a)
  n <- da[1]
  if (n != da[2])
    stop("argument 'a' is not a square matrix")
  storage.mode(a) <- "double"

  if (!is.vector(b))
    b <- as.vector(b)
  if (!is.numeric(b))
    stop("argument 'b' is not a numeric vector")
  if (n != length(b))
    stop("argument 'a' and 'b' are not compatible")

  if (m <= 0)
    stop("argument 'm' must be positive")

  mat <- matrix(0, nrow = n, ncol = m)
  storage.mode(mat) <- "double"
  z <- .C("krylov_mat",
          a = a,
          lda = as.integer(n),
          n = as.integer(n),
          b = as.double(b),
          m = as.integer(m),
          mat = mat,
          ldmat = as.integer(n),
          info = as.integer(0))
  if (z$info)
    stop(paste("krylov_mat gave error code", z$info))
  z <- z$mat
  z
}

File: pkg/src/krylov.c
/* $ID: krylov, last updated 2024-09-03, F.Osorio */

#include "fastmatrix.h"

void
krylov_mat(double *a, int *lda, int *n, double *b, int *m, double *k, int *ldk, int *info)
{ /* construcs the Krylov matrix based on an n-by-n matrix a and an n-vector b */
  FM_krylov_mat(a, *lda, *n, b, *m, k, *ldk, info);
}

void
FM_krylov_mat(double *a, int lda, int n, double *b, int m, double *k, int ldk, int *info)
{ /* given an n-by-n matrix a and an n-vector b, this function constructs the Krylov
   * matrix k, where k = [b, Ab, ..., A^(m-1)b] */
  double *work;

  /* test the input parameters */
  *info = 0;
  if (n < 0) {
    *info = -3;
  } else if (lda < MAX(1, n)) {
    *info = -2;
  } else if (m < 0) {
    *info = -5;
  } else if (ldk < MAX(1, n)) {
    *info = -7;
  }
  if (*info != 0) return;

  /* quick return if possible */
  if ((n == 0) || (m == 0))
    return;

  work = (double *) R_Calloc(n, double);
  Memcpy(work, b, n);

  /* 1st column */
  Memcpy(k, work, n);
  k += ldk;

  /* next columns */
  for (int j = 1; j < m; j++) {
    FM_mult_mat(work, a, lda, n, n, work, n, n, 1);
    Memcpy(k, work, n);
    k += ldk;
  }

  R_Free(work);
}
